Add `default-case` rule

omh1280 opened the pull request:
For #5456. Add [`default-case`](http://eslint.org/docs/rules/default-case) ESLint rule.

For every case, I've suppressed the ESLint warnings. This is because there is no good "default" case for these enums, unless throwing an error is a good default.

mramato commented:
If we're constantly suppressing them, then there's a good chance we don't need/want this rule. However, after looking at the code, I think we want to throw a `RuntimeError` in most cases rather than suppressing them. (Some may be `DeveloperError` if they are switching directly on a function argument data)

For example, `GeometryPipeline` would do something like:

```javascript
//>>includeStart('debug', pragmas.debug);
default:
    RuntimeError('Unknown GeometryType: ' + geometry.geometryType);
//>>includeEnd('debug');
```

omh1280 commented:
This reveals a DeveloperError in 3D tiles. @lilleyse mind taking a look at the Spec to see why?

omh1280 commented:
Should `var mapStyle = BingMapsStyle.AERIAL;` simply become `var mapStyle = BingMapsStyle.AERIAL_WITH_LABELS;`?

lilleyse commented:
I fixed the eslint error in `Expression.js`. Was that the only issue?

omh1280 commented:
Nope -- I didn't even see that, and there's an issue right now that travis isn't running ESLint correctly. Look at the specs in b52ac35: one fails

mramato commented:
> Should var mapStyle = BingMapsStyle.AERIAL; simply become var mapStyle = BingMapsStyle.AERIAL_WITH_LABELS;?

I'm not sure what you mean.  Can you provide some context.

omh1280 commented:
Nevermind--I think I overzealously threw a DeveloperError where it wasn't necessary. Will update soon.

omh1280 commented:
Bump!

mramato commented on Source/Renderer/Context.js:
```diff
@@ -769,6 +769,7 @@ define([
                 case gl.FRAMEBUFFER_UNSUPPORTED:
                     message = 'Framebuffer is not complete.  Unsupported: the combination of internal formats of the attached images violates an implementation-dependent set of restrictions.';
                     break;
+                // no default
```
The default here should be `message = 'Framebuffer is not complete. Unknown error condition.'`.

mramato commented on Source/Core/GeometryPipeline.js:
```diff
@@ -2546,6 +2546,10 @@ define([
                 splitLongitudeLines(instance);
                 break;
+            default:
+                //>>includeStart('debug', pragmas.debug);
```
To save a few characters in the released version, I would by the `default:` inside of the pragma when we have them.

mramato commented on Source/Scene/Model.js:
```diff
@@ -3284,6 +3284,10 @@ define([
                                 setUniforms[uniformVariableTranslate] = true;
                                 break;
+                            default:
+                                //>>includeStart('debug', pragmas.debug);
```
RuntimeErrors should not have pragmas.

mramato replied:
Same throughout, I won't comment on each one.

mramato commented on Specs/Scene/BingMapsImageryProviderSpec.js:
```diff
@@ -19,6 +20,7 @@ defineSuite([
         BingMapsImageryProvider,
         DefaultProxy,
         defined,
+        DeveloperError,
```
This isn't used in this file.

mramato commented:
Just those comments.  Once this is ready, I'd like @lilleyse to do a pass just to make sure the webgl-oriented errors we're throwing are OK.

lilleyse commented:
webgl-related changes look fine to me.

pjcozzi commented:
If we're sure about these changes, add unit tests and doc for all the new exceptions.

If this didn't catch any important bugs, I would bail on this as overkill.

mramato commented:
Providing a default case statement has long been a recommended defensive programming practice (in CS in general).  It helps catch unexpected valid and invalid values.  In STK Components we went as far as having a special `UnhandledCaseException` just for this because it was a common pattern.  I think it's worth having the rule in our projects.  

We were already following it most places and some of the changes here just take an implied default case (throwing after the switch) into an actual default case (which throws in the default). 

If you're still not convinced I can try and dig up some links to best practices.

pjcozzi commented:
If you guys really believe in this one, please go for it.

omh1280 commented:
@mramato I added one test and changed some files around to be more clear. I'm not certain about adding any `RuntimeErrors`. I think going forward this is a good practice, but I bet some functions (like those in `modelMaterialsCommon.js`) rely on returning `undefined` if no case applies.

Should we keep the `RuntimeError`s?

mramato commented:
> Should we keep the RuntimeErrors?

Sorry I didn't see this earlier, I'll take a close look at everything and get back to you.

mramato commented:
@omh1280 thanks again for trying this out.  I agree that the rule is good and if we were starting with a fresh code-base we would definitely enable it, but after looking at the changes, I'm not sure it's worth it for Cesium itself at this time.  I'm not 100% sure of some of the changes and in other cases we were already handling an implied default (throwing after a switch if no return.

We may still turn this on in other Cesium ecosystem projects on a case-by-case basis (pun intended!)

Thanks again.
